torch/nn/modules/module.py

```diff
@@ -193,13 +193,48 @@ def _apply(self, fn):
         for module in self.children():
             module._apply(fn)
 
-        for param in self._parameters.values():
+        def compute_should_change_tensor_inplace(tensor, tensor_applied):
+            if tensor.device != tensor_applied.device:
+                # If the new tensor is on a different device, then we take
+                # `torch.__future__.change_nn_module_params_inplace_cpu_cuda`
+                # into account only if we are moving the model between CPU and CUDA.
+                # Otherwise, we always replace the existing tensor.
+                if (tensor.is_cuda and tensor_applied.device == torch.device('cpu')) or \
+                   (tensor.device == torch.device('cpu') and tensor_applied.is_cuda):
+                    return torch.__future__.change_nn_module_params_inplace_cpu_cuda
+                else:
+                    return False
+            else:
+                # If the new tensor is on the same device, we always in-place
+                # update the existing tensor.
+                return True
+
+        for key, param in self._parameters.items():
             if param is not None:
-                # Tensors stored in modules are graph leaves, and we don't
-                # want to create copy nodes, so we have to unpack the data.
-                param.data = fn(param.data)
-                if param._grad is not None:
-                    param._grad.data = fn(param._grad.data)
+                param_applied = fn(param)
```

Comment:
@gchanan I tried to change this to the following:
```python
with torch.no_grad():
    print("before: param._version: ", param._version)
    param_applied = fn(param)
    print("after: param._version: ", param._version)
```
It prints:
```
before: param._version:  1
after: param._version:  1
```
I believe the reason is that `fn()` is never an in-place operation, so it doesn't bump the version counter of `param`.

Reply:
I think the reason we have `param_applied` is to avoid computing `fn(param)` twice (first in `compute_should_change_tensor_inplace(param, fn(param))` to decide on which branch to take, and then in `param.data = fn(param)` / `self._parameters[key] = Parameter(fn(param), param.requires_grad)` to set the parameter value). If we feel that having `param_applied` in this way makes the logic overly complex, we can probably accept the cost of computing `fn(param)` twice and have the two branches be more separated.

Reply:
but if you aren't inplace modifying the variable you don't need to increase the version counter.  I guess you technically don't know whether you are inplace modifying the variable because arbitrary functions can be passed in here, but again, the version counter will be correct as long as you you aren't using `.data`.

Reply:
what are you actually trying to accomplish here?  Is it that you want the "overwrite" behavior when the flag is false, but you want the exact old behavior when the flag in true?

Reply:
Note to self: we should keep the old behavior exactly the same, i.e. calling `param.data = fn(param.data)`, and for the new behavior we should call `fn(param)` and do the parameter assignment.